watchtower/wtclient/session_queue.go

```diff
@@ -0,0 +1,512 @@
+	sendInit bool) (bool, error) {
+
+	// If this is the first message being sent to the tower, we must send an
+	// Init message to establish that server supports the features we
+	// require.
+	if sendInit {
+		// Send Init to tower.
+		err := q.cfg.SendMessage(conn, q.localInit)
+		if err != nil {
+			return false, err
+		}
+
+		// Receive Init from tower.
+		remoteMsg, err := q.cfg.ReadMessage(conn)
+		if err != nil {
+			return false, err
+		}
+
+		remoteInit, ok := remoteMsg.(*wtwire.Init)
+		if !ok {
+			return true, err
+		}
+
+		// Validate Init.
+		err = q.localInit.CheckRemoteInit(
+			remoteInit, wtwire.FeatureNames,
+		)
+		if err != nil {
+			return true, err
+		}
+	}
+
+	// Send StateUpdate to tower.
+	err := q.cfg.SendMessage(conn, stateUpdate)
+	if err != nil {
+		return false, err
+	}
+
+	// Receive StateUpdate from tower.
+	remoteMsg, err := q.cfg.ReadMessage(conn)
+	if err != nil {
+		return false, err
+	}
+
+	stateUpdateReply, ok := remoteMsg.(*wtwire.StateUpdateReply)
+	if !ok {
+		return true, err
+	}
+
+	// Process the reply from the tower.
+	switch stateUpdateReply.Code {
+
+	// The tower reported a successful update, validate the response and
+	// record the last applied returned.
+	case wtwire.CodeOK:
+		lastApplied := stateUpdateReply.LastApplied
+		if lastApplied > stateUpdate.SeqNum {
+			// TODO(conner): borked watchtower?
+		}
+
+		if lastApplied < q.cfg.ClientSession.TowerLastApplied {
+			// TODO(conner): borked watchtower?
+		}
+
+		// TODO(conner): store last applied for session
+		// TODO(conner): delete blob locally
+		q.queueCond.L.Lock()
+		q.cfg.ClientSession.SeqNum++
+		q.cfg.ClientSession.TowerLastApplied = lastApplied
+		q.queue.Remove(q.queue.Front())
+		q.queueCond.L.Unlock()
+
+		return false, nil
+
+	// TODO(conner): handle other error cases properly.
+	case wtwire.StateUpdateCodeClientBehind:
+		panic("received StateUpdateCodeClientBehind from tower")
+
+	case wtwire.StateUpdateCodeMaxUpdatesExceeded:
+		panic("received StateUpdateCodeMaxUpdatexExceed from tower")
+
+	case wtwire.StateUpdateCodeSeqNumOutOfOrder:
+		panic("received StateupdateCodeSeqNumOutOfOrder from tower")
+
+	default:
+		return true, fmt.Errorf("received unknown error code %d",
+			stateUpdateReply.Code)
+	}
+}
+
+// reserveStatus returns a reserveStatus indicating whether or not the
+// sessionQueue can accept another task. reserveAvailable is returned when a
+// task can be accepted, and reserveExhausted is returned if the all slots in
+// the session have been allocated.
+//
+// NOTE: This method MUST be called with queueCond's exclusive lock held.
+func (q *sessionQueue) reserveStatus() reserveStatus {
+	seqNum := q.cfg.ClientSession.SeqNum
+	numPending := uint16(q.queue.Len())
+	maxUpdates := q.cfg.ClientSession.Policy.MaxUpdates
+
+	log.Debugf("SessionQueue %s reserveStatus seqnum=%d pending=%d "+
+		"max-updates=%d", q.ID(), seqNum, numPending, maxUpdates)
+
+	if seqNum+numPending < maxUpdates {
```

Comment:
Here's the spot I'm referring to above, re underutilization: if we have 10 spots left, and we need 15 updates, then we'll make a new session to pack those 15, rather than using the remaining 10 in this session, and making a new session for the remaining 5 updates.

Reply:
hmm, the latter is how this behaves in practice. maybe this will clear it up

`seqNum` is the highest allocated sequence number, and `numPending` is the number of bound tasks that are in the `pendingQueue` that have not been assigned a sequence number. since the updates are assigned sequentially (and not in batch), this method returns true up until the last allocated sequence number and number of unallocated updates is exactly `maxUpdates`

Reply:
Ahh ok, thanks for the explanation, was under the impression that they could be dispatched in batch.